tidy3d/components/simulation.py

```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
+            bounds[1]
+        ):
+
+            length = len(coord)
+
+            # if data does not cover structure at all take the closest index
+            if smax < coord[0]:  # structure is completely on the left side
+
+                # take 2 if possible, so that linear iterpolation is possible
+                comp_inds = np.arange(0, max(2, length))
+                is_full_cover = False
+
+            elif smin > coord[-1]:  # structure is completely on the right side
+
+                # take 2 if possible, so that linear iterpolation is possible
+                comp_inds = np.arange(min(0, length - 2), length)
+                is_full_cover = False
+
+            else:
+                if smin < coord[0]:
+                    is_full_cover = False
+                    ind_min = 0
+                else:
+                    ind_min = max(0, (coord >= smin).argmax().data - 1)
+
+                if smax > coord[-1]:
+                    is_full_cover = False
+                    ind_max = length - 1
+                else:
+                    ind_max = (coord >= smax).argmax().data
+
+                comp_inds = np.arange(ind_min, ind_max + 1)
+
+            inds_list.append(comp_inds)
+
+        return is_full_cover, data.isel(x=inds_list[0], y=inds_list[1], z=inds_list[2])
+
+    def apply_perturbations(
+        self,
+        temperature: SpatialDataArray = None,
+        electron_density: SpatialDataArray = None,
+        hole_density: SpatialDataArray = None,
+    ) -> Simulation:
+        """Apply heat and/or charge data to perturbation mediums contained in the Simulation
+        replacing them by corresponding spatially dependent custom mediums. Any of temperature,
+        electron_density, and hole_density can be 'None'. All provided fields must have identical
+        coords.
+
+        Parameters
+        ----------
+        temperature : SpatialDataArray = None
+            Temperature field data.
+        electron_density : SpatialDataArray = None
+            Electron density field data.
+        hole_density : SpatialDataArray = None
+            Hole density field data.
+
+        Returns
+        -------
+        Simulation
+            Simulation after application of heat and/or charge data.
+        """
+
+        sim_dict = self.dict()
+        structures = self.structures
+        sim_bounds = self.bounds_pml
+
+        # For each structure that contains HeatSpecSolid sample temperature field and obtain
+        # perturbed medium as a CustomMedium. Pack it into a regular Structure.
+        # Currently we perform separate interpolation inside the bounding box of each HeatStructure,
+        # a more efficient approach could be interpolating once on the entire grid and then
+        # selecting subsets of points corresponding to each structure
+        new_structures = []
+        for s_ind, s in enumerate(structures):
+            med = s.medium
+            if isinstance(med, AbstractPerturbationMedium):
```

Comment:
could this be factored into a `Structure.apply_perturbations` method? to simplify?

And for the `Simulation.medium` part, call `Medium.apply_perturbations`?

Reply:
it could be, but we would need to pass simulation bounds `sim_bounds` to it and method would have to return two values: new structure + flag whether coverage was full. So, maybe it fine to keep it all exposed

Reply:
ok